uvicorn/server.py

```diff
@@ -68,22 +78,62 @@ async def serve(self, sockets: Optional[List[socket.socket]] = None) -> None:
 
         self.lifespan = config.lifespan_class(config)
 
-        self.install_signal_handlers()
-
         message = "Started server process [%d]"
         color_message = "Started server process [" + click.style("%d", fg="cyan") + "]"
         logger.info(message, process_id, extra={"color_message": color_message})
 
-        await self.startup(sockets=sockets)
+        await self.startup(sockets=self._sockets)
         if self.should_exit:
             return
         await self.main_loop()
-        await self.shutdown(sockets=sockets)
+        await self._shutdown(sockets=self._sockets)
 
         message = "Finished server process [%d]"
         color_message = "Finished server process [" + click.style("%d", fg="cyan") + "]"
         logger.info(message, process_id, extra={"color_message": color_message})
 
+    async def start_serving(self) -> None:
+        """
+        Starts the server running in a background task and blocks until startup
+        is either fully complete, or fails.
+
+        Idempotent.  Can be called multiple times without creating multiple
+        instances.
+        """
+        if self._startup_event is None:
+            # We defer creating the startup event until start serving is called
+            # because there is no guarantee that the constructor will be called
+            # with an active loop.
+            self._startup_event = asyncio.Event()
+
+        if self._main_task is None:
+            self._main_task = asyncio.create_task(self._main())
+
+        # If the main task exits before the startup event is set it means that
+        # startup has failed.
+        await asyncio.wait(
+            [asyncio.create_task(self._startup_event.wait()), self._main_task],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+
+    async def serve(self, sockets: Optional[List[socket.socket]] = None) -> None:
+        if self._main_task is not None:
+            raise RuntimeError("cannot call serve on running server")
+
+        if sockets is not None and self._sockets is not None:
+            raise RuntimeError("cannot override already provided sockets list")
+
+        self._sockets = sockets
```

Comment:
I think we need to override `self._sockets` only in the case when `sockets=...` was passed here? Currently we would be overriding with `self._sockets = None` in most cases.

```suggestion
        if sockets is not None:
            if self._sockets is not None:
                raise RuntimeError("cannot override already provided sockets list")
            self._sockets = sockets
```

Reply:
Good catch.